modules/core/src/main/java/org/apache/ignite/internal/processors/datastructures/GridCacheLockImpl2Unfair.java

```diff
@@ -0,0 +1,784 @@
+    @Override public IgniteInClosure<GridCacheIdMessage> getReleaser() {
+        return releaser;
+    }
+
+    /** Message from previous node which release the lock. */
+    public static final class ReleasedMessage extends GridCacheIdMessage {
+        /** */
+        private static final long serialVersionUID = 181741851451L;
+
+        /** Message index. */
+        public static final int CACHE_MSG_IDX = nextIndexId();
+
+        /** Lock name. */
+        String name;
+
+        /**
+         * Empty constructor required for {@link Externalizable}.
+         */
+        public ReleasedMessage() {
+            // No-op.
+        }
+
+        /**
+         * @param id cache id.
+         * @param name lock name.
+         */
+        public ReleasedMessage(int id, String name) {
+            this.name = name;
+
+            cacheId(id);
+        }
+
+        /** {@inheritDoc} */
+        @Override public boolean addDeploymentInfo() {
+            return false;
+        }
+
+        /** {@inheritDoc} */
+        @Override public short directType() {
+            return -62;
+        }
+
+        /** {@inheritDoc} */
+        @Override public int lookupIndex() {
+            return CACHE_MSG_IDX;
+        }
+
+        /** {@inheritDoc} */
+        @Override public byte fieldsCount() {
+            return (byte)(super.fieldsCount() + 1);
+        }
+
+        /** {@inheritDoc} */
+        @Override public boolean writeTo(ByteBuffer buf, MessageWriter writer) {
+            writer.setBuffer(buf);
+
+            if (!super.writeTo(buf, writer))
+                return false;
+
+            if (!writer.isHeaderWritten()) {
+                if (!writer.writeHeader(directType(), fieldsCount()))
+                    return false;
+
+                writer.onHeaderWritten();
+            }
+
+            switch (writer.state()) {
+                case 3:
+                    if (!writer.writeString("name", name))
+                        return false;
+
+                    writer.incrementState();
+            }
+
+            return true;
+        }
+
+        /** {@inheritDoc} */
+        @Override public boolean readFrom(ByteBuffer buf, MessageReader reader) {
+            reader.setBuffer(buf);
+
+            if (!reader.beforeMessageRead())
+                return false;
+
+            if (!super.readFrom(buf, reader))
+                return false;
+
+            switch (reader.state()) {
+                case 3:
+                    name = reader.readString("name");
+
+                    if (!reader.isLastRead())
+                        return false;
+
+                    reader.incrementState();
+            }
+
+            return reader.afterMessageRead(GridCacheIdMessage.class);
+        }
+
+        /** {@inheritDoc} */
+        @Override public String toString() {
+            return S.toString(ReleasedMessage.class, this);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public void close() {
+        // No-op.
```

Comment:
why u haven't implemented lock removal ? When lock got closed its state is removed from cache, u also need to check whether somebody holds the lock while u removing it.

Reply:
I think this is a task for the other issue in Jira:-)

Reply:
mb, throw unsupported exception? Imagine, user calls close(), and expects lock closed but it doesn't close.

Reply:
I think it's a bad idea.

Reply:
why?